### PR TITLE
Add XGBoostTreeRegressor.

### PR DESCRIPTION
**Pull request by RishabhGarg108**
This PR adds the xgboost tree regressor algorithm.

### CODE REVIEW

**Comment by jonpsy** [comment]
Looks exciting! Would love to see its comparison with the original XGBoost written in C++. Documentation of the speed and correctness-metric would be nice as well. Although, I guess that was already planned, in which case, good luck!

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+               const double alpha = 0,
+               const bool warmStart = false,
+               DimensionSelectionType dimensionSelector =
+                   DimensionSelectionType());
+
+  template<typename VecType>
+  double Predict(const VecType& point) const;
+
+  template<typename MatType>
+  arma::rowvec Predict(const MatType& data) const;
```
Comment (rcurtin) [inline src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:98]: I think the usual signature of `Predict()` is `void Predict(const MatType& data, arma::rowvec& predictions);`. :+1:

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+               const double minChildWeight = 1,
+               const double gamma = 0,
+               const double lambda = 1,
+               const double alpha = 0,
+               const bool warmStart = false,
+               DimensionSelectionType dimensionSelector =
+                   DimensionSelectionType());
+
+  template<typename VecType>
+  double Predict(const VecType& point) const;
```
Comment (rcurtin) [inline src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:95]: Small comment, but this should return `typename VecType::elem_type`, which may not always be `double`. :)

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+ *   publisher = {ACM},
+ *   address = {New York, NY, USA},
+ *   keywords = {large-scale machine learning},
+ * }
+ * @endcode
+ */
+template<typename LossFunction = SSELoss,
+         typename DimensionSelectionType =
+             mlpack::tree::MultipleRandomDimensionSelect,
+         template <typename> class NumericSplitType = XGBExactNumericSplit>
```
Comment (rcurtin) [inline src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:49]: Should we allow specifying the categorical split type here?  Below I see you are always using `AllCategoricalSplit`, which I *think* will work, but it could be nice to have it user-specifiable too.

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+   * of trees.  Predict() will throw an exception until Train() is called.
+   */
+  XGBoostTreeRegressor();
+
+  /**
+   * Construct the xgboost forest and train.
+   */
+  template<typename MatType, typename ResponsesType>
+  XGBoostTreeRegressor(const MatType& dataset,
+                       const ResponsesType& responses,
```
Comment (rcurtin) [inline src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:68]: If you want to support categoricals, then I think all you would need to do is add another overload that takes a `data::DatasetInfo&` too.  I guess `SSELoss` would need some minor adaptation for that case, with a "regular" `Evaluate()` implementation.  (I think that should be pretty easy.)

**Review by rcurtin — COMMENTED** [review]
I took a closer look---the user interface looks good to me.  I hope the couple comments I left are helpful. :)  Looking forward to seeing the implementation come together! :+1:

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp (hunk 3 above)
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:49]: Actually, I did think about it. I have made the API considering that we will be having support for categorical features too in the future. When we add it in future, all we will need is to add one constructor, one public Train overload and a template parameter for it.

As of now, I didn't add a template parameter for it because we do not support it yet and it will cause unnecessary confusion.

File: src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp (hunk 4 above)
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp:68]: To be honest, I haven't given proper thought to it. I think we will also require a splitting strategy too that calculates and the sum of gradients and hessians accordingly (not 100% sure though). So, I would suggest we can leave it at this point. As I said in the previous comment, it won't be a big deal as I have considered it already.

One more thing, I haven't yet looked at how xgboost is doing it. They still have many open issues and PRs on it which I haven't had the chance to go through, but I would say we can at least take a look at them before just to be sure we are not overlooking something.

Let me know what you think. :)

**Comment by RishabhGarg108** [comment]
> Looks exciting! Would love to see its comparison with the original XGBoost written in C++. Documentation of the speed and correctness-metric would be nice as well. Although, I guess that was already planned, in which case, good luck!

Thanks! I am optimistic that our implementation will be competitive enough, but let's see how it goes. Don't forget that xgboost does use many hardware-specific optimizations :)

**Comment by rcurtin** [comment]
@RishabhGarg108 just to check in on this one, we need #3022 merged first, right?

**Comment by RishabhGarg108** [comment]
Hey @rcurtin, yes we first need to get that one merged. Then after that, a little modification will be required here to accommodate the changes made there.
